conduit/pipeline.go

```diff
@@ -208,49 +214,55 @@ func (p *pipelineImpl) Stop() error {
 	if err := (*p.exporter).Close(); err != nil {
 		p.logger.Errorf("Pipeline.Stop(): Exporter (%s) error on close: %v", (*p.exporter).Metadata().Name(), err)
 	}
-
-	return nil
 }
 
-// RunPipeline pushes block data through the pipeline
-func (p *pipelineImpl) RunPipeline() error {
+// runPipeline pushes block data through the pipeline
+func (p *pipelineImpl) runPipeline() {
+	p.running.Add(1)
+	p.err = nil
+	defer p.running.Done()
 	for {
+	pipelineRun:
 		select {
 		case <-p.ctx.Done():
-			return nil
+			return
 		default:
 			{
-				// TODO Retries?
 				p.logger.Infof("Pipeline round: %v", p.round)
 				// fetch block
 				blkData, err := (*p.importer).GetBlock(uint64(p.round))
 				if err != nil {
-					p.logger.Errorf("%v\n", err)
-					return err
+					p.logger.Errorf("%v", err)
+					p.err = err
+					goto pipelineRun
```

Comment:
this will retry forever?

Reply:
I've tried to mimic the behaviour that we have in fetcher. https://github.com/algorand/indexer/blob/develop/fetcher/fetcher.go#L219

This will run until the context is expires, which is probably whenever the cancel func is called. Errors can be raised by individual plugins, at which point we will set `p.err` so that it can be referenced by the `p.Error()` function, and the pipeline will start running again from the Importer on the same round.

Reply:
I guess this was done in order to match the current Indexer behavior?

Does Indexer also retry on processor and writer errors?

Thinking out loud here, do you think that the error handling policy should be configurable per plugin?

Reply:
> I guess this was done in order to match the current Indexer behavior?

Yes, I've tried to match the behaviour of the current Indexer here.

> Does Indexer also retry on processor and writer errors?

The Indexer's data pipeline is controlled by fetcher which runs two things concurrently:
* Fetching and enqueuing blocks
* Running the block handler on the queued blocks

The fetcher will stop running when:
* The cancel function is called or the context is `Done`
* It receives block bytes from algod which it cannot decode into a Block object
* The handler returns an error

However, the [handler we're using](https://github.com/algorand/indexer/blob/develop/cmd/algorand-indexer/daemon.go#L492) will just continue retrying when it errors. The effect is that fetcher only exits on malformed block data or context canceling. 


> Thinking out loud here, do you think that the error handling policy should be configurable per plugin?

I think it's a good idea to allow plugins to either return an error that can be retried or one that causes the pipeline to terminate. At the moment though, if we want to be backwards compatible w/ Indexer, all errors in the existing plugins would be non-fatal (except for block decoding).

Reply:
Great, thanks for the clarifications. This approach sounds good to me.